ts/config.ts

```diff
@@ -73,11 +93,116 @@ export const nody_greeter: nody_config = {
     theme_dir:
       process.env.NODY_GREETER_THEME_DIR || "/usr/share/web-greeter/themes/",
   },
+  theme: {
+    primary_html: "index.html",
+    secondary_html: "",
+  },
 };
 
 const path_to_config =
   process.env.NODY_GREETER_CONFIG || "/etc/lightdm/web-greeter.yml";
 
+let theme_dir: string | undefined;
+
+export function load_theme_dir(): string {
+  const theme = nody_greeter.config.greeter.theme;
+  const dir = nody_greeter.app.theme_dir;
+  const def_theme = "gruvbox";
+  theme_dir = path.join(dir, theme);
+
+  if (theme.startsWith("/")) theme_dir = theme;
+  else if (theme.includes(".") || theme.includes("/"))
+    theme_dir = path.join(process.cwd(), theme);
+
+  if (theme_dir.endsWith(".html")) theme_dir = path.dirname(theme_dir);
+
+  if (!fs.existsSync(theme_dir)) {
+    logger.warn(`"${theme}" theme does not exists. Using "${def_theme}" theme`);
+    theme_dir = path.join(dir, def_theme);
+  }
+
+  return theme_dir;
+}
+
+export function load_primary_theme_path(): string {
+  if (!theme_dir) load_theme_dir();
+  const dir = nody_greeter.app.theme_dir;
+  const def_theme = "gruvbox";
+  const primary = nody_greeter.theme.primary_html;
+  let path_to_theme = path.join(theme_dir, primary);
+
+  if (!path_to_theme.endsWith(".html"))
+    path_to_theme = path.join(path_to_theme, "index.html");
+
+  if (!fs.existsSync(path_to_theme)) {
+    logger.warn(
+      `"${path_to_theme}" theme does not exists. Using "${def_theme}" theme`
+    );
+    path_to_theme = path.join(dir, def_theme, "index.html");
+  }
+
+  nody_greeter.config.greeter.theme = path_to_theme;
+  return path_to_theme;
+}
+export function load_secondary_theme_path(): string {
+  if (!theme_dir) load_theme_dir();
+  const primary = nody_greeter.theme.primary_html;
+  const secondary = nody_greeter.theme.secondary_html;
+  let path_to_theme = path.join(theme_dir, secondary);
+
+  if (!path_to_theme.endsWith(".html"))
+    path_to_theme = path.join(path_to_theme, "index.html");
+
+  if (!fs.existsSync(path_to_theme)) {
+    logger.warn(
+      `"${secondary}" does not exists. Using "${primary}" for secondary monitors`
+    );
+    path_to_theme = load_primary_theme_path();
+  }
+
+  return path_to_theme;
+}
+
+export function load_theme_config(): void {
+  if (!theme_dir) load_theme_dir();
+  const path_to_theme_config = path.join(theme_dir, "index.theme");
+  try {
+    if (!fs.existsSync(path_to_theme_config))
+      throw new Error("Theme config not found");
+
+    const file = fs.readFileSync(path_to_theme_config, "utf-8");
+    const theme_config = yaml.load(file) as theme_config;
```

Comment:
Using `as` here means that we're making assumptions about the structure of a theme's / user's config file, which could lead to type errors later if they haven't done it correctly, it would be better to check that the structure of the value is of the type we expect.

There's a really cool library I've used in a number of production systems that allows you to define the types in combination with a runtime type-checker, called [`io-ts`](https://github.com/gcanti/io-ts/blob/master/index.md)

Then we could use it something like this:

```ts
import * as t from 'io-ts';

const THEME_CONFIG = t.type({
  /**
   * HTML file to use in main monitor
   * @example primary_html: "index.html"
   */
  primary_html: t.string;
  /**
   * HTML file to use in non-main (secondary) monitors
   * If the file does not exists or it's not set, `primary_html` will be used
   * @example secondary_html: "secondary.html"
   * @example secondary_html: ""
   */
  secondary_html: t.union([t.string, t.undefined]);
});

/// ...

const theme_config = yaml.load(file);
if (!THEME_CONFIG.is(theme_config)) {
  throw new Error('Invalid config').
}
```

We can also extend this functionality to automatically display an error message to the user indicating in what way their config is incorrect, using the built-in functionality of the library.

Reply:
As a slight caveat to this, if we want `secondary_html` to be an optional parameter, we probably need to do something like this instead:

```ts
const THEME_CONFIG = t.intersection([
  t.type({
    /**
     * HTML file to use in main monitor
     * @example primary_html: "index.html"
     */
    primary_html: t.string,
  }),
  t.partial({
    /**
     * HTML file to use in non-main (secondary) monitors
     * If the file does not exists or it's not set, `primary_html` will be used
     * @example secondary_html: "secondary.html"
     * @example secondary_html: ""
     */
    secondary_html: t.string,
  })
]);
```

As a parameter existing with the value undefined is different to it not existing at all, and `.is()` will say it doesn't match...